### ISSUE TITLE
Add support for traffic light and stop sign routing

### ISSUE THREAD

**Issue by TheAdventurer64** [organicmaps/organicmaps#2010]
This may already be in the works, but would it be possible to add routing support for traffic lights and stop signs? This could work by the router saying for example, "At the traffic light, make the next left.".

**Comment by biodranik** [comment]
They are very often not mapped. But we plan to add them anyway for better estimation.
@vng do we already include traffic lights and stop signs in our data?

**Comment by vng** [comment]
No, in plans to add into routing algorithm. But should we make voice instructions for traffic lights/signs ?

**Comment by biodranik** [comment]
When we have it in data, it would be easy to use it everywhere, including voice instructions, if necessary.

**Comment by vng** [comment]
I don't plan to _store_ it in data. Just to calculate graph edge's weight considering traffic lights, bumps. etc ...

**Comment by biodranik** [comment]
It could be useful to display bump icons in navigation mode and to allow adding of traffic lights and bumps from the Editor (need to check the scheme first though).

**Comment by biodranik** [comment]
Displaying traffic lights and stop sign icons while navigating is also useful.

**Comment by Zero3** [comment]
I have quite some experience with car routing, and can provide some insight with regard to a reasonable time penalty for the routing algorithm.

Long story short, **a flat penalty of 20 seconds per `highway=traffic_signals` node** is a good and simple approach that will significantly improve route selection and route time estimation compared to not handling traffic signals at all.

The longer story is that traffic signals act in many different ways in practice. Some take a short time, some take a long time, some show green more often and for a longer time depending on the direction you going, some automatically switch to green when you approach and there is no conflicting traffic, ... So to avoid overcomplicating things, it is reasonable to just use a flat penalty that matches the average waiting time over time.

Another popular app, OsmAnd, uses such a penalty. For a long time, however, they used 30 seconds as the penalty, which led to overestimation and suboptimal routes, which a number of issue reports confirmed. I did some research into this, including doing a number traces of driving in Denmark and double checking with traces by another person in another city. I inspected the traces and calculated the actual time taken on average, and ended up with 20 seconds being a very good estimate. This was passed to the OsmAnd development team, and the penalty was subsequently changed from 30 to 20 seconds.

So I thus suggest adding a flat penalty of 20 seconds per traffic signal node. Remember to take the `traffic_signals:direction` tag into account though!

If you have any questions about this, let me know! I also have a lot of insight into other car routing related things besides traffic signals. Feel free to drain my brain to improve routing and time estimation in Organic Maps.

**Comment by biodranik** [comment]
What about traffic calming devices?

**Comment by IronMaltese** [comment]
@Zero3 how should penalties work for stop signs aswell?

**Comment by euf** [comment]
I believe adding penalties for traffic calming and stop signs should definitely improve routing. I'm not sure about the exact in seconds, maybe OsmAnd defaults could serve as a good starting point. 
https://github.com/osmandapp/OsmAnd-resources/blob/master/routing/routing.xml#L896 

Furthermore, it would be great showing traffic_signals, stop, traffic_calming, crossing=uncontrolled and give_way on the map while routing. Apple Maps place them along the route. For OM, they could be just added to the `vehicle` theme (though I'm not sure if it's possible to control their Z index so they're drawn on top of the route). If it can be done, I'm ready to draw SVGs for that purpose. 

![D739FD40-F89A-4432-98D1-A9C0EC5BBF2F](https://user-images.githubusercontent.com/11754264/204104845-57d80597-2848-429c-a733-e6aa2d09c3b1.jpeg)

**Comment by Jean-BaptisteC** [comment]
https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Atraffic%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4048

**Comment by biodranik** [comment]
It may make sense to store these POIs specially, in the routing graph, instead of the normal map features. @vng WDYT?

**Comment by Zero3** [comment]
> What about traffic calming devices?

> I believe adding penalties for traffic calming and stop signs should definitely improve routing. I'm not sure about the exact values in seconds, maybe OsmAnd defaults could serve as a good starting point.

Yes and no. This probably depends a bit more on where in the world you are. Based on my experience in Denmark, they are not nearly as important as traffic signals. The OSM `traffic_calming` tag covers a wide variety of different kinds of traffic calming (see https://wiki.openstreetmap.org/wiki/Key:traffic_calming), and not all of them cause any significant delay under normal circumstances. Many traffic calming devices are actually designed to cause minimum slowdown if one drives no faster than the speed limit. There is also the issue of many traffic calming devices not being mapped in practice, so a too large penalty will nudge the routing algorithm away from well mapped roads onto less well mapped roads with unmapped traffic calming.

That being said, I think there is still value in adding a minor penalty. Most traffic calming should not cause any more delay than ~5 seconds (assuming little traffic etc.), so i would suggest starting out with a simple flat penalty of 5 seconds for all `traffic_calming` tags on nodes and ways alike.

Stepping away from the actual time to drive through these for a moment, and looking at user experience, I personally find it super annoying to drive on a road with a lot of traffic calming *bumps*. So I would add an optional feature for avoiding these by adding a synthetic penalty of ~15 seconds for `traffic_calming=bump" tags on nodes and ways.

> @Zero3 how should penalties work for stop signs aswell?

I would suggest adding a penalty of 10 seconds to `highway=stop` nodes and 5 seconds to `highway=give_way` nodes. Most of this time loss is caused by the necessary slowing down and speeding up before/after these, and not so much the actual time it takes to look for other traffic when stopped/slowed down.

(As a side note, be careful about blindly reusing values from OsmAnd. There are many issues with its routing configuration, and it is not coherent at all, but rather seems like a patchwork of many small changes and additions over time without a coherent approach and proper validation.)

**Comment by euf** [comment]
> The OSM `traffic_calming` tag covers a wide variety of different kinds of traffic calming (see https://wiki.openstreetmap.org/wiki/Key:traffic_calming), and not all of them cause any significant delay under normal circumstances. Many traffic calming devices are actually designed to cause minimum slowdown if one drives no faster than the speed limit

After examining OSM wiki and [taginfo](https://taginfo.openstreetmap.org/keys/traffic_calming#values) on the subject, I got a feeling that the following `traffic_calming=` values are actually supposed to force a driver to slow down and should be taken into account: `yes,bump,mini_bumps,hump,table,cushion,dip`.

**Comment by biodranik** [comment]
@Zero3 now there are no _turn_ penalties in OM. How would you implement them, also taking into account traffic lights, stop signs, and bumps? E.g. if there is a "give way" sign and then you need to turn left or right, etc.

**Comment by euf** [comment]
It seems only fair to add a small penalty for regular turns (e.g. 2 seconds), sharp turns and U-turns (e.g. 4 seconds). First of all, it would correspond to slowing down while performing a turn. Secondly, it should also provide simpler paths for Manhattan-style city planning: in the example below the red path would be better than the yellow one, and the blue one would be the least preferable. 

![](https://upload.wikimedia.org/wikipedia/commons/thumb/0/08/Manhattan_distance.svg/480px-Manhattan_distance.svg.png)

**Comment by Zero3** [comment]
> @Zero3 now there are no turn penalties in OM. How would you implement them, also taking into account traffic lights, stop signs, and bumps? E.g. if there is a "give way" sign and then you need to turn left or right, etc.

@biodranik I think this is quite hard to do accurately. Just to mention a few challenges:

* Some turns are "right" turns (read: turns that do not cross traffic coming from the opposite direction) done on roads with slow speed limits and ideal visual overview of the roads, leading to minimal loss of time from the turn (a "best case"). Others are "left" turns done on high speed roads with poor visual overview, leading to significant loss of time from both speeding down, waiting for opposing traffic to pass and orientation (a "worst case").
* Delays due to traffic is a big challenge when it comes to "left" turns. But that is a whole other beast, so let us just assume no traffic for now :).
* Larger intersections are particularly tricky:
  * There are many weird variants, and even more ways to map them in OSM.
  * Traffic signals are often mapped incorrectly (a common mapping mistake is placing omnidirectional `highway=traffic_signals` nodes such that you drive through multiple of them when crossing an intersection, despite only one of them affecting you in the real world).
  * Turn lanes are often not tagged (with `turn=...` or `turn:lanes=...`), and have to be inferred from the angle of the route, the road class (`highway=tertiary_link`, ...) and/or similar heuristics. OSM ways that carry both straight ahead traffic and turn traffic complicates this further.

That being said, turns take a significant amount of time when driving in cities, so I think it makes sense to at least apply a conservative penalty to them (like @euf is saying above too) to improve routing. Both in terms of actual time taken, but also to achieve simpler routes (the more turns, the more "complex" route). I believe starting out with a simple flat penalty of **10 seconds** would be a safe conservative value. 15 is probably OK too. This value could later be made to be adjusted dynamically based on turn direction, road speed limit, turning angle, ...

After that, I would look into handling some of the cases of undesirable stacking of penalties in intersections (and at stop/give way signals, like you mention as well). The most obvious case to handle (at least in Denmark, but this probably varies by region) would be to avoid applying both a traffic signal penalty and a turn penalty right after each other when "right" turning, and when "left" turning _using a turn lane_ (since the traffic lights then most likely gives you exclusive access to the road while turning in this particular scenario). But keep in mind that there is no logical mapping of intersections etc. in OSM, so it would probably have to be handled in a heuristical way. One approach could be to merge "overlapping" penalties, for examply by absorbing a 10 second turn penalty right after a 20 second traffic signal penalty.

Regarding traffic calming: I do not believe they are particularly relevant to turning penalties. Sure, there will be some corner cases with these around the world, but I believe it is better overall to keep things simple and see these are independent concepts that do not affect each other in any meaningful way.

**Comment by theadventurer62** [comment]
Regarding this, I don't believe that penalties necessarily have to be added as of right now. The voice command in and of itself would be enough to compete with other apps that do the same (eg. Google Maps, Apple, etc); Since OSM has the data, we might as well use it!

**Comment by auge8472** [comment]
> Regarding this, I don't believe that penalties necessarily have to be added as of right now. The voice command in and of itself would be enough ...

What does the voice commands solve regarding to greatly underestimated traveling time?

**Comment by TheAdventurer64** [comment]
> > Regarding this, I don't believe that penalties necessarily have to be added as of right now. The voice command in and of itself would be enough ...
> 
> What does the voice commands solve regarding to greatly underestimated traveling time?

It doesn’t. It would just be helpful for people to know when to turn.

**Comment by toas-koas** [comment]
**1. More realistic time estimation:**  
Current routing calculations often underestimate travel time because they don’t account for delays at traffic lights or stop signs. Adding fixed time penalties (e.g., 20 seconds per traffic light, 5-10 seconds per stop sign) results in much more accurate ETA, which is especially important in cities with frequent intersections.

**2. Improved route selection:**  
The algorithm will steer users toward routes with fewer stops and delays, avoiding tedious and time-consuming segments. This is particularly useful for daily commuters and anyone wanting to reach their destination with minimal interruptions.

**3. Greater user satisfaction:**  
When Organic Maps predicts true travel times and suggests realistic routes, users experience fewer surprises and maintain more trust in the app for planning.

**Comment by jpc-ae** [comment]
Any movement on this? I was suprised this wasn't implimented yet considering the age of the project, but I do understand the non-trivial nature of dealing with complex intersections. I would still rather a slightly off penalty be applied than none at all.
